# 이미지 인식 - DNN

In [22]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPooling2D
from tensorflow.keras.callbacks import ModelCheckpoint, EarlyStopping

import os
import matplotlib.pyplot as plt
%matplotlib inline

In [23]:
# seed 값 설정
seed = 0
np.random.seed(seed)
tf.random.set_seed(seed)

In [74]:
# 데이터 불러오기
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()
X_train = X_train.reshape(X_train.shape[0], 784).astype('float32') / 255 
X_test = X_test.reshape(X_test.shape[0], 784).astype('float32') / 255 
Y_train = tf.keras.utils.to_categorical(Y_train)
Y_test = tf.keras.utils.to_categorical(Y_test)

In [89]:
# 컨볼루션 신경망 설정
model = Sequential([
    Dense(512, input_dim =784, activation='relu'),
    Dense(300, activation='relu'),
    Dense(119, activation='relu'),
    Dense(10, activation='softmax')
])
model.summary()

Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_44 (Dense)             (None, 512)               401920    
_________________________________________________________________
dense_45 (Dense)             (None, 300)               153900    
_________________________________________________________________
dense_46 (Dense)             (None, 119)               35819     
_________________________________________________________________
dense_47 (Dense)             (None, 10)                1200      
Total params: 592,839
Trainable params: 592,839
Non-trainable params: 0
_________________________________________________________________


In [90]:
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

In [91]:
# 모델 최적화 설정
MODEL_DIR = './model/'
if not os.path.exists(MODEL_DIR):
    os.mkdir(MODEL_DIR)

In [92]:
modelpath = MODEL_DIR + "mnist-cnn-{epoch:02d}-{val_loss:.4f}.hdf5"
checkpointer = ModelCheckpoint(filepath=modelpath, monitor='val_loss', 
                               verbose=1, save_best_only=True)
early_stopping_callback = EarlyStopping(monitor='val_loss', patience=20)

In [93]:
# 모델의 실행
history = model.fit(X_train, Y_train, validation_data=(X_test, Y_test), epochs=30,
                    batch_size=500, verbose=2, 
                    callbacks=[early_stopping_callback, checkpointer])

Train on 60000 samples, validate on 10000 samples
Epoch 1/30

Epoch 00001: val_loss improved from inf to 0.14909, saving model to ./model/mnist-cnn-01-0.1491.hdf5
60000/60000 - 1s - loss: 0.3607 - accuracy: 0.8964 - val_loss: 0.1491 - val_accuracy: 0.9554
Epoch 2/30

Epoch 00002: val_loss improved from 0.14909 to 0.09789, saving model to ./model/mnist-cnn-02-0.0979.hdf5
60000/60000 - 1s - loss: 0.1189 - accuracy: 0.9643 - val_loss: 0.0979 - val_accuracy: 0.9690
Epoch 3/30

Epoch 00003: val_loss improved from 0.09789 to 0.08872, saving model to ./model/mnist-cnn-03-0.0887.hdf5
60000/60000 - 1s - loss: 0.0740 - accuracy: 0.9771 - val_loss: 0.0887 - val_accuracy: 0.9723
Epoch 4/30

Epoch 00004: val_loss improved from 0.08872 to 0.07437, saving model to ./model/mnist-cnn-04-0.0744.hdf5
60000/60000 - 1s - loss: 0.0515 - accuracy: 0.9840 - val_loss: 0.0744 - val_accuracy: 0.9771
Epoch 5/30

Epoch 00005: val_loss improved from 0.07437 to 0.06846, saving model to ./model/mnist-cnn-05-0.0685.hd

In [94]:
from tensorflow.keras.models import load_model
del model
model = load_model('model/mnist-cnn-09-0.0607.hdf5')

In [95]:
# 테스트 정확도 출력 
print("\n Test Accuracy: %.4f" % (model.evaluate(X_test, Y_test, verbose=0)[1]))


 Test Accuracy: 0.9840


### 결론: Best Model은 Epoch 9, 정확도는 98.40%

# 이미지 인식 - CNN

In [96]:
# 데이터 불러오기
(X_train, Y_train), (X_test, Y_test) = mnist.load_data()
X_train = X_train.reshape(X_train.shape[0], 28, 28, 1).astype('float32') / 255
X_test = X_test.reshape(X_test.shape[0], 28, 28, 1).astype('float32') / 255
Y_train = tf.keras.utils.to_categorical(Y_train)
Y_test = tf.keras.utils.to_categorical(Y_test)

In [115]:
# 컨볼루션 신경망 설정
model = Sequential([
    Conv2D(32, kernel_size=(3, 3), input_shape=(28, 28, 1), activation='relu'),
    Conv2D(64, (3, 3), activation='relu'),
    MaxPooling2D(pool_size=2),
    Dropout(0.5),
    Flatten(),
    Dense(512, activation='relu'),
    Dense(256, activation='relu'),
    Dropout(0.5),
    Dense(10, activation='softmax')
])
model.summary()

Model: "sequential_14"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_4 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_5 (Conv2D)            (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 12, 12, 64)        0         
_________________________________________________________________
dropout_4 (Dropout)          (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_2 (Flatten)          (None, 9216)              0         
_________________________________________________________________
dense_53 (Dense)             (None, 512)               4719104   
_________________________________________________________________
dense_54 (Dense)             (None, 256)             

In [116]:
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

In [117]:
# 모델 최적화 설정
MODEL_DIR = './model2/'
if not os.path.exists(MODEL_DIR):
    os.mkdir(MODEL_DIR)

In [118]:
modelpath = MODEL_DIR + "mnist-cnn-{epoch:02d}-{val_loss:.4f}.hdf5"
checkpointer = ModelCheckpoint(filepath=modelpath, monitor='val_loss', 
                               verbose=1, save_best_only=True)
early_stopping_callback = EarlyStopping(monitor='val_loss', patience=10)

In [119]:
# 모델의 실행
history = model.fit(X_train, Y_train, validation_data=(X_test, Y_test), epochs=40,
                    batch_size=500, verbose=2, 
                    callbacks=[early_stopping_callback, checkpointer])

Train on 60000 samples, validate on 10000 samples
Epoch 1/40

Epoch 00001: val_loss improved from inf to 0.05754, saving model to ./model2/mnist-cnn-01-0.0575.hdf5
60000/60000 - 46s - loss: 0.3203 - accuracy: 0.9000 - val_loss: 0.0575 - val_accuracy: 0.9826
Epoch 2/40

Epoch 00002: val_loss improved from 0.05754 to 0.03884, saving model to ./model2/mnist-cnn-02-0.0388.hdf5
60000/60000 - 45s - loss: 0.0761 - accuracy: 0.9774 - val_loss: 0.0388 - val_accuracy: 0.9870
Epoch 3/40

Epoch 00003: val_loss improved from 0.03884 to 0.03349, saving model to ./model2/mnist-cnn-03-0.0335.hdf5
60000/60000 - 45s - loss: 0.0509 - accuracy: 0.9849 - val_loss: 0.0335 - val_accuracy: 0.9892
Epoch 4/40

Epoch 00004: val_loss improved from 0.03349 to 0.02782, saving model to ./model2/mnist-cnn-04-0.0278.hdf5
60000/60000 - 45s - loss: 0.0382 - accuracy: 0.9882 - val_loss: 0.0278 - val_accuracy: 0.9908
Epoch 5/40

Epoch 00005: val_loss did not improve from 0.02782
60000/60000 - 45s - loss: 0.0313 - accuracy

In [120]:
from tensorflow.keras.models import load_model
del model
model = load_model('model2/mnist-cnn-12-0.0241.hdf5')

In [121]:
# 테스트 정확도 출력 
print("\n Test Accuracy: %.4f" % (model.evaluate(X_test, Y_test, verbose=0)[1]))


 Test Accuracy: 0.9929


### 결론: Best Model은 Epoch 11, 정확도는 99.29%